src/pybind/mgr/volumes/fs/subvolume.py

```diff
@@ -141,6 +141,46 @@ def remove_subvolume(self, spec, force):
         except cephfs.Error as e:
             raise VolumeException(-e.args[0], e.args[1])
 
+    def resize_subvolume(self, spec, newsize, extend=True):
+        """
+        :param spec: subvolume path specification
+        :param newsize: new size In bytes
+        :param extend: the extend/shrink switch
+        :return: None
+        """
+
+        if newsize <= 0:
+            raise VolumeException(-errno.EINVAL, "Provide a valid size")
+
+        subvolpath = spec.subvolume_path
+
+        try:
+            maxbytes = int(self.fs.getxattr(subvolpath, 'ceph.quota.max_bytes').decode('utf-8'))
+        except cephfs.NoData:
+            raise VolumeException(-errno.EINVAL, "ceph.quota.max_bytes: No such attribute")
+
+        try:
+            usedbytes = int(self.fs.getxattr(subvolpath, "ceph.dir.rbytes").decode('utf-8'))
+        except cephfs.NoData:
+            raise VolumeException(-errno.EINVAL, "ceph.dir.rbytes: No such attribute")
+
+        if extend:
+            if newsize <= maxbytes:
+                raise VolumeException(-errno.EINVAL, "Can't extend the subvolume. "
+                                      "The new size must be greater than the current size '{0}'".format(maxbytes))
+        else:
+            if newsize >= maxbytes:
+                raise VolumeException(-errno.EINVAL, "Can't shrink the subvolume. "
+                                      "The new size must be lesser than the current size '{0}'".format(maxbytes))
+            elif newsize < usedbytes:
+                raise VolumeException(-errno.EINVAL, "Can't shrink the subvolume. "
+                                      "The new size cannot be lesser than the current used size '{0}', "
+                                      "possible data loss.".format(usedbytes))
```

Comment:
I don't see how data loss is possible? I also think it's valuable to reduce the quota of a subvolume after it has already used too much space: the user must free files until they are back under the quota.

Reply:
@batrick, do you see a benefit of returning a warning message letting the user know that they must free some  storage to continue writing?

Reply:
At least in manila, we raise an error when used size > shrunk size.
https://github.com/openstack/manila/commit/9ff0d1312#diff-4256e8eb2400d0325fe689bc36010201R278 

@tombarron , do we really need to do this ? Is this in the manila API spec? If so we can introduce a `ceph fs subvolume get usedsize` API, or something similar like the one in ceph_volume_client.

Reply:
> @batrick, do you see a benefit of returning a warning message letting the user know that they must free some storage to continue writing?

A warning is good, sure. Maybe there should be a message in any case that gives the current usage, the quota, and the % used.

Reply:
With some Manila back ends the size is not just a quota and there is danger of data loss if a share is shrunk less than the amount actually used so Manila evolved with the expectation that back ends will not allow shrinking below current usage.  But we can enforce this in the Manila CephFS driver as ajarr suggests with the ceph mgr interface if we have a way to get the usedsize.

Reply:
Thanks @tombarron 

@batrick @ajarr So what do you finally suggest for the resize? Do we need a check and except in `resize_subvolume()`? Based on that I could write the tests @batrick had suggested.

Reply:
As we discussed during standup, the plan is to have an option like `--no-shrink` which will accommodate Manila.

Reply:
> As we discussed during standup, the plan is to have an option like `--no-shrink` which will accommodate Manila.
Thanks.  Probably the option should have a different name though -- we allow shrinking a share, just not below current usage.